starlite/datastructures/headers.py

```diff
@@ -1,15 +1,396 @@
+  this list of conditions and the following disclaimer in the documentation
+  and/or other materials provided with the distribution.
+
+* Neither the name of the copyright holder nor the names of its
+  contributors may be used to endorse or promote products derived from
+  this software without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+"""
+
 import re
 from abc import ABC, abstractmethod
-from typing import Any, ClassVar, Dict, Optional
-
+from typing import (
+    TYPE_CHECKING,
+    Any,
+    ClassVar,
+    Dict,
+    Iterable,
+    Iterator,
+    List,
+    Mapping,
+    Optional,
+    Tuple,
+    Union,
+)
+
+from multidict import CIMultiDict, CIMultiDictProxy
 from pydantic import BaseModel, Extra, Field, ValidationError, validator
 from typing_extensions import Annotated
 
 from starlite.exceptions import ImproperlyConfiguredException
+from starlite.utils import deprecated
+
+if TYPE_CHECKING:
+    from starlite.types.asgi_types import Message, RawHeadersList, Scope
 
 ETAG_RE = re.compile(r'([Ww]/)?"(.+)"')
 
 
+def _encode_headers(headers: Iterable[Tuple[str, str]]) -> "RawHeadersList":
+    return [(key.lower().encode("latin-1"), value.encode("latin-1")) for key, value in headers]
+
+
+class Headers(CIMultiDictProxy[str]):
+    """An immutable, case-insensitive [multidict](https://multidict.aio-
+    libs.org/en/stable/multidict.html#cimultidictproxy) for HTTP headers."""
+
+    def __init__(self, headers: Optional[Union[Mapping[str, str], "RawHeadersList"]] = None) -> None:
+        headers_: Union[Mapping[str, str], List[Tuple[str, str]]] = {}
+        if isinstance(headers, list):
+            headers_ = [(key.decode("latin-1"), value.decode("latin-1")) for key, value in headers]
+        elif headers:
+            headers_ = headers
+
+        super().__init__(CIMultiDict(headers_))
+
+    @classmethod
+    def from_scope(cls, scope: "Scope") -> "Headers":
+        """Create headers from a `Scope`.
+
+        Args:
+            scope: An ASGI Scope
+
+        Returns:
+            Headers
+        """
+        return cls(list(scope["headers"]))
+
+    @classmethod
+    def from_message(cls, message: "Message") -> "Headers":
```

Comment:
`Message` is too wide here, it should be restrictred only to the ASGI event types that support headers IMO.